tfsdk/attribute.go

```diff
@@ -477,9 +478,10 @@ func (a Attribute) validateAttributes(ctx context.Context, req ValidateAttribute
 }
 
 // modifyPlan runs all AttributePlanModifiers
-func (a Attribute) modifyPlan(ctx context.Context, req ModifyAttributePlanRequest, resp *ModifyAttributePlanResponse) {
+func (a Attribute) modifyPlan(ctx context.Context, req ModifyAttributePlanRequest, resp *ModifyAttributePlanResponse, schemaResp *ModifySchemaPlanResponse) {
```

Comment:
Two things on this:

1. Could we avoid having the double responses by just refactoring this to be at the schema level, instead of at the attribute level? Meaning it gets a `ModifySchemaPlanRequest`, a `*tftypes.AttributePath` and a `*ModifySchemaPlanResponse` and just updates them for that attribute? That would avoid the double bookkeeping, right?
2. I think in other places we have decided that the children should have their plan modifiers run before the parents, e.g. the elements before the list, so that the list could assume the plan for the elements was accurate when its plan modifier ran. This looks like it's running on the parent first, then on the elements. Is there a reason it needs to be this way, or can we talk about doing the children first, then the parent?

Reply:
1. We can probably drop the `ModifyAttributePlanResponse` double bookkeeping, I think that should just work. 👍  For the request, if we're going to need to pass extra information (path) to be able to use `ModifySchemaPlanRequest`, I think sticking with `ModifyAttributePlanRequest` which already has that field or creating a separate type might be best.
2. If I'm reading the current logic correct, it does the current attribute plan modifiers before children: https://github.com/hashicorp/terraform-plugin-framework/blob/33080f200a6d9f58eb9b8a9e8e9c5e0a4483e27c/tfsdk/schema.go#L311-L343 -- if we want to change this, its probably worth thinking about the design some more because there are likely arguments either which way on which to run first.

Reply:
https://github.com/hashicorp/terraform-plugin-framework/pull/224/commits/386ac38a8a7b2449fe38a89d809d3305c976cb0e drops the double response in preference of just `ModifySchemaPlanResponse`.

Reply:
Oops! Sorry. My bad re: 2.

I do think the doing it children first, then parent makes more sense, just because then the parent can see the modified children for, e.g., comparing for equality, checking values, etc. If we do parent first, then children, we're either working with out of date children (in which case, what happens if both modify the child in conflicting ways? Who wins?) or we get the updated children, but no way to influence the parent on the results of the final child values.

I dunno if that makes sense or not, but that's my two cents.

Reply:
Created https://github.com/hashicorp/terraform-plugin-framework/issues/225 for future noodling! 🍝